helm/Idol-nifi: support additional extensions

sam-jeffrey-mf opened the pull request:
- Persist NiFi extensions in the `statedata` volume
- Copy additional extensions, specified by `nifi.additionalExtensionFiles`, into the persisted extensions directory during the `ensure-conf` init container

Allows NiFi extensions to be imported from any additional volume specified on the container, and persists any manually imported extensions, e.g. from `kubectl cp /my/extension.nar idol-nifi-0:/opt/nifi/nifi-current/extensions/`

daniel-lau-mf commented on helm/idol-nifi/values.yaml:
```diff
@@ -169,6 +169,9 @@ nifi:
     version: ""
+    
+  # -- additional NiFi extension files to be copied into the NiFi extensions directory during pod initialization
+  additionalExtensionFiles: []
```
Add to `values.schema.json`

daniel-lau-mf commented:
Is it worth us considering some form of simple script hook in the `ensure-conf` to allow sourcing of extensions from other locations (e.g. download from an S3 bucket or similar?)

sam-jeffrey-mf commented:
> Is it worth us considering some form of simple script hook in the `ensure-conf` to allow sourcing of extensions from other locations (e.g. download from an S3 bucket or similar?)

I think this is a good idea. Do you think calling a script, provided by a 'nifi.ensureResourcesScriptFile' property, at the end of the current ensure-conf script would suffice? The script reference by this property could then be mounted from a config map

daniel-lau-mf commented:
> > Is it worth us considering some form of simple script hook in the `ensure-conf` to allow sourcing of extensions from other locations (e.g. download from an S3 bucket or similar?)
> 
> I think this is a good idea. Do you think calling a script, provided by a 'nifi.ensureResourcesScriptFile' property, at the end of the current ensure-conf script would suffice? The script reference by this property could then be mounted from a config map

Yes, that would probably suffice.

Alternatively, in the IDOL containers we iterate through scripts mounted in a `prestart_scripts` directory:
```
run_prestart_scripts() {
  for script in /${IDOL_COMPONENT}/prestart_scripts/*.sh; do [ -f "$script" ] && source "$script"; done
}
```
Though admittedly that's not done in an init container stage (so more like a step in the nifi container startup - maybe we should add that into the base nifi images).

sam-jeffrey-mf commented:
> Alternatively, in the IDOL containers we iterate through scripts mounted in a `prestart_scripts` directory:
> 
> ```
> run_prestart_scripts() {
>   for script in /${IDOL_COMPONENT}/prestart_scripts/*.sh; do [ -f "$script" ] && source "$script"; done
> }
> ```
> 
> Though admittedly that's not done in an init container stage (so more like a step in the nifi container startup - maybe we should add that into the base nifi images).

Assuming that this would run prior to NiFi being started in the main image, I think this would be a good idea

sam-jeffrey-mf commented:
I have removed all references to `additionalExtensionFiles` in favor of using a prestart scripts. I have added a startup probe, with configurable failure time (defaulting to 1 min) to allow time for the prestart scripts to run, and added a section to the readme to document how additional extensions should be provided.

daniel-lau-mf commented on helm/idol-nifi/templates/nifi/nifi.yaml:
```diff
@@ -55,14 +55,19 @@ spec:
+            cp -rv /opt/nifi/nifi-current/conf/ /mnt/store/nifi/
           fi
-
+          echo Copying files into extensions directory
+          cp -rnv /opt/nifi/nifi-current/extensions/ /mnt/store/nifi/
```
Do we need this in `ensure-conf` now?

sam-jeffrey-mf replied:
From our off ticket conversation:
> definitely still want persisting extensions directory 
> that at least lets you write a prestart script that checks if your required extensions are present before downloading?
> downloading/copying

I think that it is still useful to persist this directory, for the reasons you stated above, because it also maintains any NARs manually copied into that directory, and because it would be difficult for the end user to achieve the same behavior because providing their own mount for `/opt/nifi/nifi-current/extensions` would overwrite the originally provided NARs.
